### PR TITLE
Wrap socket binding with a loop to handle port-in-use errors

### PR DESCRIPTION
**Pull request by shane-kerr**
tl;dr Re-try socket binding if the source address is in use, as this indicates we picked a bad socket.

More words:

The code has an error condition where if a port is already in use an exception is raised. This is problematic because dnsruby picks a random port on its own. There are many reasons why a port may be in use. For example, another application could be using it, or a client may want to use lots of simultaneous queries and happen to pick the same port twice. This appears to hit quite often in TCP sessions, which keep ports unusable for longer than they are in use. (Note that we could reduce the problem in the case of TCP by setting the SO_REUSEADDR flag, but we will still have other reasons why a port may be in use.)

The only place that can reliably find an unused port is the network stack. dnsruby rejects letting the OS pick the port, with the following comment:
```ruby
      # Apparently, Linux will just give you the same port as last time, unless it is still
      # open, in which case you get n+1.
```
This does not appear to be true with modern versions of Linux (at least Debian Stable using a kernel from 2012 picks random ports), but presumably this was the behavior was at some point, and surely people are using other OS which may have this property.

Since we must pick our own ports sometimes, we need code which does this without failing.

The basic approach to this has to be a retry on the EADDRINUSE error. There are several heuristics that we could use around this technique.

The approach in this patch is the most basic one, where we simply try again if the address that we picked is in use. There is a a theoretical failure case here where we actually have all of the ports unavailable. In that case we'll spin forever. This is not a very realistic issue, but it is possible. A more likely scenario is that _most_ of the ports are unavailable, in which case picking one at random will take a lot of extra CPU to find one that is acceptable.

We could work around this in several ways. The easiest would be to add a counter and have a maximum number of attempts to get a port. Alternately we could do the shuffling technique mentioned in the port selection code, so that we only try each port once (or perhaps N times).

I've opted to leave addressing this problem as an exercise for future work, should it ever become a problem for anyone. ;)

### CODE REVIEW

**Comment by alexdalitz** [comment]
Hi Shane - 

Nice to hear from you! Thanks for helping with dnsruby.

I have to say that it's been a long day (not to mention a long week!) so please excuse any complete idiocy in this post. Disclaimer aside...

Can I please ask what version of dnsruby you are using?

I'm a little confused by your post, as you imply that Dnsruby will raise an exception if it cannot bind to a socket. I don't believe that this is the case. AFAIUI, dnsruby will only raise an exception if one of the following occurs (once we have failed to bind to the socket) -

o Errno::EBADF, Errno::ENETUNREACH
o the client has specifically requested that we use a defined port and we can't
o we've tried more than 50 ports

Do you find that this is not happening? If so, could you please provide a test case?

Do you have a problem with the current behaviour, if it is as I outline above?

Thanks again for your input,


Alex.

**Comment by shane-kerr** [comment]
Alex,

On Thu, 22 Jan 2015 10:22:48 -0800
Alex Dalitz <notifications@github.com> wrote:

> Can I please ask what version of dnsruby you are using?

This is a fresh clone from the master on GitHub.

> I'm a little confused by your post, as you imply that Dnsruby will
> raise an exception if it cannot bind to a socket. I don't believe
> that this is the case. AFAIUI, dnsruby will only raise an exception
> if one of the following occurs (once we have failed to bind to the
> socket) -
> 
> o Errno::EBADF, Errno::ENETUNREACH
> o the client has specifically requested that we use a defined port
> and we can't o we've tried more than 50 ports
> 
> Do you find that this is not happening? If so, could you please
> provide a test case?

An easy way to cause the behavior is to change get_next_src_port to
always return the same port:

------------------------------------------------------------------------
diff --git a/lib/dnsruby/packet_sender.rb b/lib/dnsruby/packet_sender.rb
index 9bc9446..a177b01 100644
--- a/lib/dnsruby/packet_sender.rb
+++ b/lib/dnsruby/packet_sender.rb
@@ -491,6 +491,7 @@ module Dnsruby
 
 
     def get_next_src_port
+      return 40000
       # Different OSes have different interpretations of "random port"
here. # Apparently, Linux will just give you the same port as last
time, unless it is still # open, in which case you get n+1.
------------------------------------------------------------------------

If you run the online tests ("rake test_online") you'll see the errors
in the TCP tests:

  1) Error:
TestTcp#test_bad_truncation:
IOError: dnsruby can't connect to 127.0.0.1:39375 from 0.0.0.0:40000,
  use_tcp=true, exception = Errno::EADDRINUSE, Address already in use -
  bind(2) for "0.0.0.0" port
  40000 /home/shane/dyn/dnsruby/lib/dnsruby/resolver.rb:248:in
  `send_message' /home/shane/dyn/dnsruby/lib/dnsruby/resolver.rb:195:in
  `query' /home/shane/dyn/dnsruby/test/tc_tcp.rb:181:in
  `test_bad_truncation'


The real code hits the condition much less often, of course, since it
picks randomly from 15535 ports. However, with a lot of parallel
connections it happens more often. (And as I noted in the original pull
request, it is always possible that some other application is using any
given port, and there is no way for dnsruby to know this.)

I think that with N simultaneous ports in use, your chance of getting a
collision is something like this:

    P(failure) = 1 - (15534/15535)^N

In our case, we run 100 TCP queries in parallel, which has a 0.7%
chance of collision. However, since the TCP sessions take several
minutes to timeout because the SO_REUSEADDR flag is not set, we
actually end up needing over 800 ports, giving us nearly a 5% chance of
failure on ever run.


It would be tricky to write a unit test for this, but not impossible.
Probably the most straightforward way is to bind() a given TCP port,and
then use a custom version of get_next_src_port method which returns
that port number first and then a different port number. The first call
to get_next_src_port would result in the Errno::EADDRINUSE exception,
and the second would not.

I decided not to bother with writing a test, since the code I
submitted is clearly perfect in every way. ;)


Cheers,

--
Shane

**Comment by alexdalitz** [comment]
Proper response later but for now please note that your solution behaves just the same as the current code if get source port always returns same unavailable port. Except yours loops forever...

I still fail to see the problem case for doing anything other than raising max number of retries from 50?

Thanks,

Alex.

**Comment by shane-kerr** [comment]
Alex,

On Fri, 23 Jan 2015 07:37:11 -0800
Alex Dalitz <notifications@github.com> wrote:

> Proper response later but for now please note that your solution
> behaves just the same as the current code if get source port always
> returns same unavailable port. Except yours loops forever...

Well, yes, the PR loops forever if you give it an in-use port.

My point, however, was just to demonstrate the condition where
Errno::EADDRINUSE was raised, since you didn't seem to think that it
was possible.

I agree there is an infinite loop if you somehow don't have any free
ports. I mentioned this in the pull request:

    "There is a a theoretical failure case here where we actually have
    all of the ports unavailable. In that case we'll spin forever. This
    is not a very realistic issue, but it is possible."

> I still fail to see the problem case for doing anything other than
> raising max number of retries from 50?

I also discuss possible mitigations for the infinite loop, including the
one that you mention here of having a maximum number of retries.

    "We could work around this in several ways. The easiest would be to
    add a counter and have a maximum number of attempts to get a port.
    Alternately we could do the shuffling technique mentioned in the
    port selection code, so that we only try each port once (or perhaps
    N times)."

The main problem with having a counter is that you will sometimes end up
raising an exception occasionally when there _are_ available ports. For
example, if you only have 50 available ports then you'll still fail to
get one 85% of the time with 50 iterations. (Note that on average
you'll have to loop 215 times to get a port with the code I submitted -
hardly ideal either.)

The suggestion of using a shuffling technique is basically to shuffle
the 15535 ports, and try them in-order. This is a relatively expensive
operation (shuffling being as expensive as sorting, I think), and can
still fail if (for example) sockets are being created and destroyed by
another application dynamically.

Probably the "ultimate" solution (in terms of balancing the desire to
always get a port but not wanting to waste time) would be something like
this:

1. See if you are running on a system that properly randomizes source
   port. In this case, use 0, and let the OS pick a source port.

2. Otherwise, try a number of iterations (like your 50).

3. If that fails, then shuffle the available ports, and walk through
   them, trying to find an available port.

4. If that fails, let the Errno::EADDRINUSE through.


Unfortunately Linux doesn't tell you whether it supports port
randomization, but the commit that added that was made in 2005, which
means that it is in all currently supported kernels:

https://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/commit/net/ipv4/inet_connection_sock.c?id=6df716340da3a6fdd33d73d7ed4c6f7590ca1c42

There are likely a couple of RHEL users who are still running kernels
older than that... RHEL 4 doesn't EOL until early 2017. Personally I
wouldn't worry about it and would just assume that Linux is fine.
(People running 10+ year old kernels are not likely to be dropping a
lot of new software onto their systems.)

I think the only problem OS these days is likely to be OS X, and maybe
Windows (it's hard to tell with Windows):

https://www.cymru.com/jtk/misc/ephemeralports.html

I guess to be safe I would whitelist the "known good" OS.

Anyway, this seems like a lot of work, which is why I submitted the
straightforward patch. :) If you want I can put in a counter, or even
implement the "full" solution. Alternately, I can just write this up as
a comment in the port selection code.

Really, infinite loops won't ever be a problem in actual usage....

Cheers,

--
Shane

**Comment by alexdalitz** [comment]
Hi Shane - 

Thanks for your detailed messages!

I suspect that we are in violent agreement, but speaking past each other. You say -

"My point, however, was just to demonstrate the condition where
Errno::EADDRINUSE was raised, since you didn't seem to think that it
was possible."

This is absolutely not the case!  What I said was that dnsruby would give up IF IT GOT 50 Errno::EADDRINUSE IN A ROW.

Dnsruby will retry getting a port 50 times before it gives up.

It seems to me that you do not think that this is the case - your comments appear to suggest that you think that dnsruby will give up as soon as it fails to get a port.

Could you please confirm that you are aware of the current retry mechanism?

If so, could you please explain why your infinite retry loop is better than the current behaviour of retrying N times (where N is currently 50)?

Thanks!


Alex.

**Comment by shane-kerr** [comment]
Alex,

On Fri, 23 Jan 2015 10:05:21 -0800
Alex Dalitz <notifications@github.com> wrote:

> I suspect that we are in violent agreement, but speaking past each
> other. You say -

WHOOPS. You are correct. :-P
 
> "My point, however, was just to demonstrate the condition where
> Errno::EADDRINUSE was raised, since you didn't seem to think that it
> was possible."
> 
> This is absolutely not the case!  What I said was that dnsruby would
> give up IF IT GOT 50 Errno::EADDRINUSE IN A ROW.
> 
> Dnsruby will retry getting a port 50 times before it gives up.
> 
> It seems to me that you do not think that this is the case - your
> comments appear to suggest that you think that dnsruby will give up
> as soon as it fails to get a port.
> 
> Could you please confirm that you are aware of the current retry
> mechanism?

I was *not* aware of the current retry mechanism. :(
 
> If so, could you please explain why your infinite retry loop is
> better than the current behaviour of retrying N times (where N is
> currently 50)?

I'm not sure that it is, but perhaps we can investigate a bit together.

We definitely saw EADDRINUSE in our production setup. (We have since
segmented our checks so that we use different source IP addresses for
each independent batch and the exceptions have disappeared.)

I'll double-check the version of dnsruby that we're using there. How
long ago was the retry introduced?


In the meantime, I did a very little investigation this morning. Using
git version 0a735887ebf450154cc3117cc089cfc2bf53fa88 I was able to get
EADDRINUSE after just 3 or 4 runs of "rake test_offline". It turns out
that this is because the "test_TCP_port" test always uses port 57923,
which means that if tests are run too quickly then the port will not be
available. I set this to a random port and the tests always work now:

diff --git a/test/tc_tcp.rb b/test/tc_tcp.rb
index 197ef8b..1c90216 100644
--- a/test/tc_tcp.rb
+++ b/test/tc_tcp.rb
@@ -27,7 +27,7 @@ class TestTcp < Minitest::Test
   def test_TCP_port
     #  Need a test server so we can tell what port this message was
actually sent on! port = nil
-    src_port = 57923
+    src_port = 50000 + rand(15535)
     Dnsruby::PacketSender.clear_caches
     received_port = nil
     server_thread = Thread.new {


(Yes, this has the same chance of failure as production code. No, that
doesn't matter.)

Possibly it is worthwhile to make this minor tweak to the tests?



Anyway, sorry for the noise. I'll get back to you as soon as the person
who built our production images is available.

Cheers,

--
Shane

**Comment by alexdalitz** [comment]
> > If so, could you please explain why your infinite retry loop is
> > better than the current behaviour of retrying N times (where N is
> > currently 50)?
> 
> I'm not sure that it is, but perhaps we can investigate a bit together.
> 
> We definitely saw EADDRINUSE in our production setup. (We have since
> segmented our checks so that we use different source IP addresses for
> each independent batch and the exceptions have disappeared.)
> 
> I'll double-check the version of dnsruby that we're using there. How
> long ago was the retry introduced?

The retry has been there for several years.

I’m happy to up the number of retries from 50 - which is a number pulled out of thin air :-)   Any suggestions?

> In the meantime, I did a very little investigation this morning. Using
> git version 0a735887ebf450154cc3117cc089cfc2bf53fa88 I was able to get
> EADDRINUSE after just 3 or 4 runs of "rake test_offline". It turns out
> that this is because the "test_TCP_port" test always uses port 57923,
> which means that if tests are run too quickly then the port will not be
> available. I set this to a random port and the tests always work now:
> 
> Possibly it is worthwhile to make this minor tweak to the tests?

Yes - thanks!

> Anyway, sorry for the noise. I'll get back to you as soon as the person
> who built our production images is available.

:-)


Alex Dalitz
alex@caerkettontech.com

**Comment by keithrbennett** [comment]
In case it's helpful, you can use ?w=1 in the Github URL to tell it to ignore whitespace in the comparison:

https://github.com/alexdalitz/dnsruby/pull/60/files?w=1?diff=split

**Comment by alexdalitz** [comment]
Hi - 
Is this still an issue for you?
Would you like me to change the number of retries from 50?
Can I close this issue yet?

Thanks!


Alex.

**Comment by shane-kerr** [comment]
Alex,

On Thu, 29 Jan 2015 23:43:28 -0800
Alex Dalitz <notifications@github.com> wrote:

> Is this still an issue for you?
> Would you like me to change the number of retries from 50?
> Can I close this issue yet?

Go ahead and close it.

Our fix (using separate IP addresses for each batch of checks) works for
us, so I haven't spent cycles digging into it further.

My suspicion is that we're just using too many TCP ports, and even with
50 we hit this condition. If I'm feeling ambitious I'll come up with a
test that duplicates this. If this is indeed the problem, what might be
a better solution is to try setting the SO_REUSEADDR flag on the
socket.

Anyway, I'll let you know if I come up with some results in that area.

Thanks, and sorry again for the noise. :(

--
Shane
